sympy/polys/matrices/sdm.py

```diff
@@ -0,0 +1,432 @@
+"""
+
+Module for the SDM class.
+
+"""
+
+from operator import add, neg, pos, sub
+from collections import defaultdict
+
+from .exceptions import DDMBadInputError, DDMDomainError
+
+from .ddm import DDM
+
+
+class SDM(dict):
+    """Sparse matrix based on polys domain elements
+
+    This is a dict subclass and is a wrapper for a dict of dicts that supports
+    basic matrix arithmetic +, -, *, **.
+    """
+    def __init__(self, elemsdict, shape, domain):
+        super().__init__(elemsdict)
+        self.shape = self.rows, self.cols = m, n = shape
+        self.domain = domain
+
+        if not all(0 <= r < m for r in self):
+            raise DDMBadInputError("Row out of range")
+        if not all(0 <= c < n for row in self.values() for c in row):
+            raise DDMBadInputError("Column out of range")
+
+    @classmethod
+    def new(cls, sdm, shape, domain):
+        return cls(sdm, shape, domain)
+
+    def copy(A):
+        Ac = {i: Ai.copy() for i, Ai in A.items()}
+        return A.new(Ac, A.shape, A.domain)
+
+    @classmethod
+    def from_list(cls, ddm, shape, domain):
+        m, n = shape
+        if not (len(ddm) == m and all(len(row) == n for row in ddm)):
+            raise DDMBadInputError("Inconsistent row-list/shape")
+        getrow = lambda i: {j:ddm[i][j] for j in range(n) if ddm[i][j]}
+        irows = ((i, getrow(i)) for i in range(m))
+        sdm = {i: row for i, row in irows if row}
+        return cls(sdm, shape, domain)
+
+    @classmethod
+    def from_ddm(cls, ddm):
+        return cls.from_list(ddm, ddm.shape, ddm.domain)
+
+    def to_list(M):
+        m, n = M.shape
+        zero = M.domain.zero
+        ddm = [[zero] * n for _ in range(m)]
```

Comment:
Could this be `[[zero]*n]*m`?

Reply:
This is a classic gotcha in Python. Creating a list of lists with double multiplication results in a list containing many copies of the same list. That's okay for an immutable structure but otherwise mutating one row alo mutates every other (because they are all the same Python list):
```python
In [1]: M = [[0] * 3] * 3                                                                                                         

In [2]: M                                                                                                                         
Out[2]: [[0, 0, 0], [0, 0, 0], [0, 0, 0]]

In [3]: M[0][0] = 2                                                                                                               

In [4]: M                                                                                                                         
Out[4]: [[2, 0, 0], [2, 0, 0], [2, 0, 0]]
```

Reply:
Yes,of course. I have made that mistake before but still did forget...